### ISSUE TITLE
isfProfile in settings/profile.json is not updated when changed from multiple isf to one

### ISSUE THREAD

**Issue by PieterGit** [openaps/oref0#708]
I changed from multiple isf to only one.  I noticed the `isfProfile` in `settings/profile.json` was still having the old multiple isf format, e.g.

```
 "isfProfile": {
    "units": "mg/dL",
    "user_preferred_units": "mmol/L",
    "sensitivities": [
      {
        "i": 0,
        "start": "00:00:00",
        "sensitivity": <<autotune_sens>>,
        "x": 0,
        "offset": 0,
        "endOffset": 420
      },
      {
        "i": 14,
        "start": "07:00:00",
        "sensitivity": <<old_sens2>>,
        "x": 1,
        "offset": 420
      },
      {
        "i": 36,
        "start": "18:00:00",
        "sensitivity": <<old_sens3>>,
        "x": 2,
        "offset": 1080,
        "endOffset": 1440
      }
    ],
    "first": 2
  },
```

The ISF profile is ok in `settings/pumpprofile.json`:
```
 "sens": <<new_sens>>,
  "isfProfile": {
    "units": "mg/dL",
    "user_preferred_units": "mmol/L",
    "sensitivities": [
      {
        "i": 0,
        "start": "00:00:00",
        "sensitivity": <<new_sens>>,
        "x": 0,
        "offset": 0,
        "endOffset": 1440
      }
    ],
    "first": 2
  },
``` 

Autotune fills in the first sensitivity entry in the `settings/profile.json`: e.g. `<<autotune_sens>>`

When I look the same applied to `settings/insulin_sensitivities.json`.
I could fix `settings/insulin_sensitivities.json` with:
``` 
openaps report invoke settings/insulin_sensitivities.json
```

I tried to fix ithe `settings/profile.json`  with the following set of commands:
```
openaps get-settings
openaps report invoke settings/insulin_sensitivities.json
openaps report invoke settings/autosens.json
openaps report invoke settings/pumpprofile.json
openaps report invoke settings/profile.json
openaps report invoke settings/insulin_sensitivities_raw.json
```
But it didn't help.  Can you please explain how I should fix this so that when the ISF profile is changed (even from multiple ISF to one) this is fixed automatically?

Tested with 0.6.0-dev, branch insulinPeakTime ( https://github.com/openaps/oref0/pull/695 ).

**Comment by PieterGit** [comment]
Today re-installed 0.6.0-dev1 and moved old directory out of the way. This seems the best temporary solution. Can others confirm that changing a multiple ISF to a single ISF is not reflected in the profile json files (even after a long wait).

Any thoughts how to find the root cause, fo why changing ISF on the pump does not seem to propagate to `settings/profile.json` and `settings/pumpprofile.json` and `settings/autotune.json`.

**Comment by CaptainBalou** [comment]
I am working on dev branch.
I can confirm that my latest profile changes are reflected in settings/pumpprofile.json but not in settings/profile.json nor settings/autotune.json. 
I also noticed that other values are not sync over all those files.

**Comment by tepidjuice** [comment]
Happy to test this. How long should the files take to update after changing from multiple ISF's to a single ISF?

**Comment by kcrcgm** [comment]
I'd love to help test this. I just changed (two days ago i think) from multiple isf and cr to a single one. How long do i wait before checking if they are synced? How do i move the old directory out of the way?

**Comment by PieterGit** [comment]
@kcrcgm : just check the files mentioned above to  see if they have multiple ISF's. You can better format them, e.g. `jq . settings/profile.json`

Pumphistory is refreshed each hour I believe and autotune files are updated at 5 past midnight. 

If they still have multiple ISF's  after that time a workaround is to:
```
/etc/init.d/cron stop
l
(check to see pump loop has ended)
cd 
mv myopenaps myopenaps.multiple_isf
bash myopenaps.multiple_isf/oref0-runagain.sh
/etc/init.d/cron start
```
Note this is only a workaround, and reinstalls oref0 in a fresh directory.  It will disable SMB until the next autotune run. We should find the root cause and make sure that a ISF or CR change is detected automatically.

**Comment by vebaba** [comment]
I have same issue. Changed from multiples ISF to single but I can see old values reported in AOPS pill. I'm on 0.6.0-dev2

**Comment by tim2000s** [comment]
I inadvertently screwed up my profiles earlier today so had to redo all the reports. It will be interesting to see the outcome of this with a single carb ratio versus the stored autotune results.

**Comment by bdr1177** [comment]
In addition to the issue above, I've found another possibly related issue.  If the pump is programmed with multiple ISF segments of the same ISF, the rig will only use the autotuned ISF from midnight until the next programmed segment.  When it reaches the next programmed ISF segment, it will switch over to using the ISF that is programmed into the pump as the baseline for making adjustments.

**Comment by tim2000s** [comment]
@bdr1177 I think it’s always done that. 

My fiddling about yesterday that resulted in a load of manual work I think has identified where the issue is. 

I don’t think the single ISF profile file is being properly copied to Autotune.json and pumpprofile.json when you change to a single isf. The night before I had changed to a single isf and saw the issue listed here in autotune. 

Yesterday I deleted profile files from settings and had to regenerate the pumpprofile.json file from the newly created profile.json. This morning after running Autotune at midnight I have a single ISF - I need to check another rig where I didn’t remove pumpprofile.json and see what the age of the file is.

**Comment by tim2000s** [comment]
Ive had a dig into this and I understand what’s going on. 

At the start of the process we copy the settings/autotune.json file over to the autotune/profile.json and then run autotune from there. 

If we’ve changed the pump ISF profile from multiple to single ISF it is never picked up as the settings/profile.json file is ignored, even if we’ve deleted the autotune generated file and pulled profile from the pump. 

I think the fix for this is to diff autotune.json and profile.json and check the time stamps of the two files, and if they are different and settings/profile.json is newer, use the settings/profile.json as the start point for running autotune. 

I inadvertently did this last night after deleting half my settings folder and it picked up the single ISF model.

**Comment by bdr1177** [comment]
@tim2000s Yes, I assume this has always been the case.  I just didn't realize that my issue was related to #708.  I had switched over to a single ISF some time ago (although it was still in the pump as several different time segments with the same ISF) and had done a complete `rm -r myopenaps` and reinstall, so I assumed the rig would interpret this as one ISF rather than multiple.  I now understand that the because of the multiple time segments, this was being interpreted as multiple ISFs and was leading to the problem described in this issue.  Sounds like it should be a relatively easy problem to fix, but unfortunately I don't have the programming ability to do so myself.  (Trying to learn in my free time, but not there yet.)

**Comment by scottleibrand** [comment]
@tim2000s The timestamp check and conditional copy you're describing is exactly what `cp -up` does. Want to try adding another one of those for these files and see if it fixes this?

**Comment by tim2000s** [comment]
I’ll give it a try.

**Comment by PieterGit** [comment]
@vebaba reports at https://gitter.im/nightscout/intend-to-bolus?at=5a2905a6a2be4668286f7e2d that:
"I just manually edited settings/autosens.json, settings/profile.json, autotune/profile.json and maybe autotune/profile.pump.json. I took ISF paragraph from settings/pumpprofile.json. Essentially, fixing was copy/paste of correct single ISF profile that was created from my pump settings... After that and one Autotune rerun, everything got correct for me"

I looked into it in the past, and couldn't find the root cause so that this will be fixed without user intervention by either refreshing the pump history or an autotune run. I couldn't find / solve it. Perhaps @scottleibrand is able to shed some light of what might be going wrong here.

**Comment by tim2000s** [comment]
I’ll add a line to autotune this weekend and test the switch over. It should fix the issue. If it does, will PR in.

**Comment by PieterGit** [comment]
@tim2000s Great!
@scottleibrand Can you target this for milestone 0.6.1?

**Comment by tim2000s** [comment]
OKay - so adding `cp -up settings/pumpprofile.json settings/autotune.json` fixes the multi ISF problem but resets autotune back to pump settings. Are we okay with this behaviour?

**Comment by endafarrell** [comment]
I can also test this: specifically, my ISF in the late afternoon has been considerably higher than other times during the day.

**Comment by PieterGit** [comment]
@tim2000s : where did you put the extra `cp -up settings/pumpprofile.json settings/autotune.json`? Perhaps we should create a `oref0-merge-profile` similar to https://github.com/openaps/oref0/blob/dev/bin/oref0-get-profile.js that will copy all the original pump settings (`settings/pumpprofile.json`?)  and merge it with the autotune settings for CSF / ISF and basals from `settings/autotune.json` ?.

Is the `settings/autotune.json`  the authoritative file for the autotuned CSF/ISF and basals? According to https://github.com/openaps/oref0/blob/master/bin/oref0-autotune-recommends-report.sh#L52-L64 the reports shows `$directory/autotune/profile.pump.json` for the pump settings and `$directory/autotune/profile.json` for the autotuned profile. @scottleibrand : can you explain what files we should merge?

**Comment by danamlewis** [comment]
Can someone update on this?

**Comment by PieterGit** [comment]
I think @jpcunningh fixed this problem with https://github.com/openaps/oref0/pull/1101 recently.
I will create a new rig soon, and will see if I can retest this with current dev and confim that it will be resolved in 0.7.x.

**Comment by jpcunningh** [comment]
@PieterGit, #1101 only maintains the DIA and insulinePeakTime, and only maintains those if useCustomPeakTime is not set. It doesn't maintain the ISF profile.

Since Autotune is trying to tune the ISF values, it may not be as simple to find the logic to keep up with pump profile changes that should reset the autotuned derived values.

**Comment by old-square-eyes** [comment]
I think I am experiencing this.

My old ISF used to be...
00:00 - 3
06:30 - 2
11:00 - 3

My new ISF is...
00:00 - 2.7
06:30 - 1.5
11:00 - 2.7

Profile still reflects the old (and so are ISF/autosens calcs, pump profile reflects the new.

**Comment by old-square-eyes** [comment]
Does anyone know how to fix it? - The workaround steps up above don't look clear to me.

**Comment by danamlewis** [comment]
I believe this is resolved in 0.7.0 now - closing, please re-open if this was not fully operationalized and we can add the workaround to 0.7.1 if it's not already there.
